Checking link to video with hash anchor results in `NotImplementedError`/`AssertionError`

Checking an mp4 video with hash anchor caused the error: `NotImplementedError: subclasses of ParserBase must override error()` with the following traceback:

```
Checking all links that haven't been tested for 10080 minutes.
Traceback (most recent call last):
  File "/opt/integreat-cms/.venv/bin/integreat-cms-cli", line 53, in <module>
    main()
  File "/opt/integreat-cms/.venv/bin/integreat-cms-cli", line 49, in main
    execute_from_command_line(sys.argv)
  File "/opt/integreat-cms/.venv/lib/python3.9/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/opt/integreat-cms/.venv/lib/python3.9/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/integreat-cms/.venv/lib/python3.9/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/integreat-cms/.venv/lib/python3.9/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/opt/integreat-cms/.venv/lib/python3.9/site-packages/linkcheck/management/commands/checklinks.py", line 35, in handle
    external_checked = check_links(external_recheck_interval=externalinterval, limit=limit, check_internal=False)
  File "/opt/integreat-cms/.venv/lib/python3.9/site-packages/linkcheck/utils.py", line 116, in check_links
    status = u.check_url(check_internal=check_internal, check_external=check_external)
  File "/opt/integreat-cms/.venv/lib/python3.9/site-packages/linkcheck/models.py", line 182, in check_url
    self._check_external(self.url, external_recheck_interval)
  File "/opt/integreat-cms/.venv/lib/python3.9/site-packages/linkcheck/models.py", line 328, in _check_external
    self._check_anchor(anchor, response.text, internal=False)
  File "/opt/integreat-cms/.venv/lib/python3.9/site-packages/linkcheck/models.py", line 337, in _check_anchor
    names = parse_anchors(html)
  File "/opt/integreat-cms/.venv/lib/python3.9/site-packages/linkcheck/__init__.py", line 98, in parse_anchors
    parser.feed(content)
  File "/usr/lib/python3.9/html/parser.py", line 110, in feed
    self.goahead(0)
  File "/usr/lib/python3.9/html/parser.py", line 178, in goahead
    k = self.parse_html_declaration(i)
  File "/usr/lib/python3.9/html/parser.py", line 263, in parse_html_declaration
    return self.parse_marked_section(i)
  File "/usr/lib/python3.9/_markupbase.py", line 149, in parse_marked_section
    sectName, j = self._scan_name( i+3, i )
  File "/usr/lib/python3.9/_markupbase.py", line 390, in _scan_name
    self.error("expected name token at %r"
  File "/usr/lib/python3.9/_markupbase.py", line 33, in error
    raise NotImplementedError(
NotImplementedError: subclasses of ParserBase must override error()
```

I could reproduce this problem with any direct video link followed by a hash anchor, e.g.:

https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ElephantsDream.mp4#t=1.0

Python: 3.9.2
Django: 3.2.16
django-linkcheck: 2.0.0

I just noticed that the ultimate `NotImplementedError` was indeed a bug of python itself: https://bugs.python.org/issue31844

On Python 3.10 and newer (https://github.com/python/cpython/commit/e34bbfd61f405eef89e8aa50672b0b25022de320), it throws an `AssertionError` instead and the traceback looks e.g. as follows:
```
Checking all links that haven't been tested for 10080 minutes.
Traceback (most recent call last):
  File "/opt/integreat-cms/.venv/bin/integreat-cms-cli", line 53, in <module>
    main()
  File "/opt/integreat-cms/.venv/bin/integreat-cms-cli", line 49, in main
    execute_from_command_line(sys.argv)
  File "/opt/integreat-cms/.venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/opt/integreat-cms/.venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/integreat-cms/.venv/lib/python3.10/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/integreat-cms/.venv/lib/python3.10/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/opt/integreat-cms/.venv/lib/python3.10/site-packages/linkcheck/management/commands/checklinks.py", line 35, in handle
    external_checked = check_links(external_recheck_interval=externalinterval, limit=limit, check_internal=False)
  File "/opt/integreat-cms/.venv/lib/python3.10/site-packages/linkcheck/utils.py", line 116, in check_links
    status = u.check_url(check_internal=check_internal, check_external=check_external)
  File "/opt/integreat-cms/.venv/lib/python3.10/site-packages/linkcheck/models.py", line 182, in check_url
    self._check_external(self.url, external_recheck_interval)
  File "/opt/integreat-cms/.venv/lib/python3.10/site-packages/linkcheck/models.py", line 328, in _check_external
    self._check_anchor(anchor, response.text, internal=False)
  File "/opt/integreat-cms/.venv/lib/python3.10/site-packages/linkcheck/models.py", line 337, in _check_anchor
    names = parse_anchors(html)
  File "/opt/integreat-cms/.venv/lib/python3.10/site-packages/linkcheck/__init__.py", line 98, in parse_anchors
    parser.feed(content)
  File "/usr/lib/python3.10/html/parser.py", line 110, in feed
    self.goahead(0)
  File "/usr/lib/python3.10/html/parser.py", line 178, in goahead
    k = self.parse_html_declaration(i)
  File "/usr/lib/python3.10/html/parser.py", line 263, in parse_html_declaration
    return self.parse_marked_section(i)
  File "/usr/lib/python3.10/_markupbase.py", line 154, in parse_marked_section
    raise AssertionError(
AssertionError: unknown status keyword 'H' in marked section
```

I managed to trace back the problem to d45fb6212c213db5877bd34fcda52f24cf600284 where I changed a bare `except` to an `except UnicodeDecodeError` which apparently did not cover all possible exceptions.

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,5 +1,7 @@
 Unreleased
 
+* Fix `NotImplementedError`/`AssertionError` when checking
+  video links with hash anchors (Timo Ludwig, #150)
 * Skip checking of hash anchors for non-html files
 * recheck/ignore/unignore requests were using an obsolete `request.is_ajax` call
   (#147)

--- a/linkcheck/models.py
+++ b/linkcheck/models.py
@@ -378,17 +378,23 @@ class Url(models.Model):
             else:
                 try:
                     names = parse_anchors(html)
+                # Known possible errors include: AssertionError, NotImplementedError, UnicodeDecodeError
+                except Exception as e:
+                    logger.debug(
+                        '%s while parsing anchors: %s',
+                        type(e).__name__,
+                        e
+                    )
+                    self.message += ', failed to parse HTML for anchor'
+                    if not TOLERATE_BROKEN_ANCHOR:
+                        self.status = False
+                else:
                     if self.anchor in names:
                         self.message += f', working {scope} hash anchor'
                     else:
                         self.message += f', broken {scope} hash anchor'
                         if not TOLERATE_BROKEN_ANCHOR:
                             self.status = False
-                except UnicodeDecodeError as e:
-                    logger.debug('UnicodeDecodeError while parsing anchors: %s', e)
-                    self.message += ', failed to parse HTML for anchor'
-                    if not TOLERATE_BROKEN_ANCHOR:
-                        self.status = False
 
 
 class Link(models.Model):

--- a/linkcheck/tests/sampleapp/views.py
+++ b/linkcheck/tests/sampleapp/views.py
@@ -38,4 +38,8 @@ def http_redirect_to_anchor(request):
 
 
 def static_video(request):
-    return HttpResponse(b"", content_type='video/mp4')
+    return HttpResponse(b'', content_type='video/mp4')
+
+
+def static_video_forged_content_type(request):
+    return HttpResponse(b'<![x02\x00\xa0\xcc', content_type='text/html')

--- a/linkcheck/tests/test_linkcheck.py
+++ b/linkcheck/tests/test_linkcheck.py
@@ -280,6 +280,12 @@ class ExternalCheckTestCase(LiveServerTestCase):
         self.assertEqual(uv.status, True)
         self.assertEqual(uv.message, "200 OK")
 
+    def test_forged_video_with_time_anchor(self):
+        uv = Url(url=f"{self.live_server_url}/static-files/fake-video.mp4#t=2.0")
+        uv.check_url()
+        self.assertEqual(uv.status, True)
+        self.assertEqual(uv.message, "200 OK, failed to parse HTML for anchor")
+
 
 class ModelTestCase(TestCase):
 

--- a/linkcheck/tests/urls.py
+++ b/linkcheck/tests/urls.py
@@ -24,4 +24,5 @@ urlpatterns = [
     path('http/anchor/', views.http_response_with_anchor),
     path('timeout/', views.timeout),
     path('static-files/video.mp4', views.static_video),
+    path('static-files/fake-video.mp4', views.static_video_forged_content_type),
 ]

